server/sonar-server/src/main/java/org/sonar/server/qualityprofile/ws/QProfileProjectsAction.java

```diff
@@ -0,0 +1,164 @@
+        @Override
+        protected Long doApply(ProjectQprofileAssociationDto input) {
+          return input.getProjectId();
+        }
+      });
+
+      final Collection<Long> authorizedProjectIds = dbClient.authorizationDao().keepAuthorizedProjectIds(session, projectIds, UserSession.get().userId(), UserRole.USER);
+      Iterable<ProjectQprofileAssociationDto> authorizedProjects = Iterables.filter(projects, new Predicate<ProjectQprofileAssociationDto>() {
+        @Override
+        public boolean apply(ProjectQprofileAssociationDto input) {
+          return authorizedProjectIds.contains(input.getProjectId());
+        }
+      });
+
+      Paging paging = Paging.create(pageSize, page, authorizedProjectIds.size());
+
+      List<ProjectQprofileAssociationDto> pagedAuthorizedProjects = Lists.newArrayList(authorizedProjects);
+      if (pagedAuthorizedProjects.size() <= paging.offset()) {
+        pagedAuthorizedProjects = Lists.newArrayList();
+      } else if (pagedAuthorizedProjects.size() > paging.pageSize()) {
+        pagedAuthorizedProjects = pagedAuthorizedProjects.subList(paging.offset(), paging.offset() + pageSize);
+      }
+
+      writeProjects(response.newJsonWriter(), pagedAuthorizedProjects, paging);
+    } finally {
+      session.close();
+    }
+  }
+
+  private void checkProfileExists(String profileKey, DbSession session) {
+    try {
+      dbClient.qualityProfileDao().getNonNullByKey(session, profileKey);
+    } catch (IllegalArgumentException doesNotExist) {
+      NotFoundException notFoundException = new NotFoundException(String.format("Could not find a quality profile with key '%s'", profileKey));
+      notFoundException.initCause(doesNotExist);
+      throw notFoundException;
+    }
+  }
+
+  private List<ProjectQprofileAssociationDto> loadProjects(String profileKey, DbSession session, String selected, String query) {
+    List<ProjectQprofileAssociationDto> projects = Lists.newArrayList();
+    if (SELECTION_SELECTED.equals(selected)) {
+      projects.addAll(dbClient.qualityProfileDao().selectSelectedProjects(profileKey, query, session));
+    } else if (SELECTION_DESELECTED.equals(selected)) {
+      projects.addAll(dbClient.qualityProfileDao().selectDeselectedProjects(profileKey, query, session));
+    } else {
+      projects.addAll(dbClient.qualityProfileDao().selectProjectAssociations(profileKey, query, session));
+    }
+    return projects;
+  }
+
+  private void writeProjects(JsonWriter json, List<ProjectQprofileAssociationDto> projects, Paging paging) {
+    json.beginObject();
+    json.name("results").beginArray();
+    for (ProjectQprofileAssociationDto project : projects) {
+      json.beginObject()
+        .prop("key", project.getProjectUuid())
+        .prop("name", project.getProjectName())
+        .prop("selected", project.isAssociated())
+        .endObject();
+    }
+    json.endArray();
+    json.prop("more", paging.hasNextPage());
```

Comment:
@stas-vilchik it would be a good idea, as I thing we should have a common way to define paging in WS.
But it's up to you web guys to validate this :)